### PR TITLE
[WIP]Lossless Transformations(Rotate feature)

### PR DESCRIPTION
**Pull request by shankarpriyank**
**Description (required)**

Made simple changes to initiate the editing feature in the app, right now the UI can only facilitate rotating the images, the rotated imaged are written in the download directory, but how to propagate the images back to the uploads is something I am trying to figure rn, and I would appreciate any help with it.

Also cause I have merged #5220 in this branch so please ignore the changes that were made in that pr.

Please let me know what do you think about the changes, I know that a lot of improvements can be made but I am focusing on getting things running first, let me know if you want me to do anything differently

### CODE REVIEW

**Review by nicolas-raoul — COMMENTED** [review]
Very nice, great start! :-)

**Comment by shankarpriyank** [comment]
@nicolas-raoul Thank you, I am having a hard time trying to find a way to send the edited back to the upload screen.
Do you have any advice/ idea on how it might be done?

**Comment by shankarpriyank** [comment]
@nicolas-raoul can you do a quick review here, I was not sure where to add the helper functions.
I am unable to upload any image right now, even from the master branch. So could not test the functionality.
I guess that now maybe that the rotated image will get uploaded(not sure though!), trying to figure out how to reflect the change in the preview of the image too.

**Comment by nicolas-raoul** [comment]
Hi @shankarpriyank sorry for the delay!

I just uploaded [this picture](https://commons.wikimedia.org/wiki/File:Building_in_Darmstadt.jpg) using this branch, rotating it and tapping "Save". Unfortunately the uploaded image is not rotated, as you can see, it is deeper than a preview issue. For testing you can just take pictures of streets or monuments ad rotate them first to an unnatural angle.

Thanks! :-)

**Comment by shankarpriyank** [comment]
No worries @nicolas-raoul, thanks for testing it, I will try finding to fix this

**Comment by shankarpriyank** [comment]
Hi @nicolas-raoul can you test this branch again?
I tested it with a bunch of images and it was working as intended.

**Comment by nicolas-raoul** [comment]
I just uploaded https://commons.wikimedia.org/wiki/File:2007,_cliff_near_Rabat_in_Malta.jpg with your branch. The picture originally had the wrong orientation, but I corrected it using the `Edit image` button and now it looks great. :-)

One remark: The result image does not have EXIF anymore. Would you mind making sure that the EXIF is kept? It is vital metadata information of the picture, it can contain coordinates and exact date for instance, which are very important to use the picture in Wikipedia.

Thanks a lot!

**Comment by shankarpriyank** [comment]
How are you checking the exif data? is there a handy tool i can use to test it ?

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]
Somehow I get `Failed to rotate image` now... I tried two different pictures, 3 times 90 degrees.

**Comment by nicolas-raoul** [comment]
https://github.com/commons-app/apps-android-commons/assets/99590/2e017171-0df7-436c-88cf-0472e3614dfe

**Comment by shankarpriyank** [comment]
That's happening most probably because the image's encoding is not supported by lljtran(mostly happens huffman encoding), can you please try with some other images.

**Comment by nicolas-raoul** [comment]
The successful rotation I had mentioned at https://github.com/commons-app/apps-android-commons/pull/5252#issuecomment-1750706215 was with a picture taken by the same camera the same day, though, os I doubt it used a different encoding...

I just tried with a picture taken by Pixel 7 Pro's stock camera, the error message does not appear but the picture is not rotated either. :'-(

**Comment by shankarpriyank** [comment]
Ahh, lemme have a look, can you email me the images you tried?

**Comment by nicolas-raoul** [comment]
Sent to your email. :-)

By the way, all pictures are correctly rotated by this app which is a GUI over LLJtran: https://play.google.com/store/apps/details?id=jp.kame.jpegcropper&pli=1 It is not open source but here is the open source equivalent written by the same person: https://github.com/kamemak/ajpegtran_example/tree/pixelization_extension

**Comment by shankarpriyank** [comment]
@nicolas-raoul I tested this, the new error is popping up cause lljtran is having [trouble](https://gist.github.com/shankarpriyank/05da1e91b64cd5eaceaefa954fe3162c) reading the exif data, and very weirdly sometimes even when there is no problem while reading the rotation is not working as intended.
I am thinking of a workaround, we store the exif data of the original image temporarily somewhere and then copy it to transformed image, but for that we will need a list of exif tags we are interested in.
by having a look at the logs my guess is that we are mostly interested in there
```
2023-10-08 21:01:20.111  4012-4012  UploadModel             fr.free.nrw.commons                  D  File created date is 1174833926000
2023-10-08 21:01:20.113  4012-4012  GGGGG                   fr.free.nrw.commons                  E  [Location]
2023-10-08 21:01:20.116  4012-4012  FileMetadataUtils       fr.free.nrw.commons                  D  Retuning tags for pref:Author
2023-10-08 21:01:20.118  4012-4012  FileProcessor           fr.free.nrw.commons                  D  Checking for tag: Artist
2023-10-08 21:01:20.120  4012-4012  FileProcessor           fr.free.nrw.commons                  D  Checking for tag: CameraOwnerName
2023-10-08 21:01:20.121  4012-4012  FileMetadataUtils       fr.free.nrw.commons                  D  Retuning tags for pref:Copyright
2023-10-08 21:01:20.122  4012-4012  FileProcessor           fr.free.nrw.commons                  D  Checking for tag: Copyright
2023-10-08 21:01:20.123  4012-4012  FileMetadataUtils       fr.free.nrw.commons                  D  Retuning tags for pref:Camera Model
2023-10-08 21:01:20.125  4012-4012  FileProcessor           fr.free.nrw.commons                  D  Checking for tag: Make
2023-10-08 21:01:20.126  4012-4012  FileProcessor           fr.free.nrw.commons                  D  Exif tag Make with value KYOCERA redacted.
2023-10-08 21:01:20.127  4012-4012  FileProcessor           fr.free.nrw.commons                  D  Checking for tag: Model
2023-10-08 21:01:20.128  4012-4012  FileProcessor           fr.free.nrw.commons                  D  Exif tag Model with value Finecam SL400R  redacted.
2023-10-08 21:01:20.129  4012-4012  FileMetadataUtils       fr.free.nrw.commons                  D  Retuning tags for pref:Lens Model
2023-10-08 21:01:20.130  4012-4012  FileProcessor           fr.free.nrw.commons                  D  Checking for tag: LensMake
2023-10-08 21:01:20.131  4012-4012  FileProcessor           fr.free.nrw.commons                  D  Checking for tag: LensModel
2023-10-08 21:01:20.132  4012-4012  FileProcessor           fr.free.nrw.commons                  D  Checking for tag: LensSpecification
2023-10-08 21:01:20.134  4012-4012  FileMetadataUtils       fr.free.nrw.commons                  D  Retuning tags for pref:Serial Numbers
2023-10-08 21:01:20.135  4012-4012  FileProcessor           fr.free.nrw.commons                  D  Checking for tag: BodySerialNumber
2023-10-08 21:01:20.136  4012-4012  FileProcessor           fr.free.nrw.commons                  D  Checking for tag: LensSerialNumber
2023-10-08 21:01:20.137  4012-4012  FileMetadataUtils       fr.free.nrw.commons                  D  Retuning tags for pref:Software
2023-10-08 21:01:20.139  4012-4012  FileProcessor           fr.free.nrw.commons                  D  Checking for tag: Software
2023-10-08 21:01:20.139  4012-4012  FileProcessor           fr.free.nrw.commons                  D  Exif tag Software with value FC-SL400R_ Ver 1.00 redacted.
2023-10-08 21:01:20.161  4012-4012  FileProcessor           fr.free.nrw.commons                  D  Calling GPSExtractor
```

**Comment by shankarpriyank** [comment]
I just now pushed a commit, which copies all the available exif data.
If you align with the approach, maybe you can test it.

**Comment by nicolas-raoul** [comment]
Tested with https://commons.wikimedia.org/wiki/File:March_2007_narrow_street_in_Valetta.jpg , both rotation and EXIF seem to go well. :-)
I will test with more pictures, hopefully it can be merged once we switch to Flamingo (which should be soon).
Meanwhile would you mind adding javadoc for all new classes and methods? Thanks!

**Comment by shankarpriyank** [comment]
I will add in the java docs soon, I would say its better to not blocked by #5220 , once the feature is fully tested I can either push in a commit reverting the extra changes, or maybe cherry pick the commits to a new branch.

**Comment by shankarpriyank** [comment]
hi @nicolas-raoul I am having a hard time deciding where to write java docs and where not(assuming that the code is self explanatory), can you please leave a comment at the places where you think a java doc would help?

**Comment by nicolas-raoul** [comment]
It looks like thumbnails of rotated pictures do not show up when uploading? Not a big problem. If not easy to solve, would you mind creating an issue about it after I merge?

File: app/src/main/java/fr/free/nrw/commons/di/EditActivityModule.kt
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/di/EditActivityModule.kt:8]: class javadoc

File: app/src/main/java/fr/free/nrw/commons/di/EditActivityModule.kt
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/di/EditActivityModule.kt:11]: function javadoc

File: app/src/main/java/fr/free/nrw/commons/edit/EditActivity.kt
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/edit/EditActivity.kt:29]: class javadoc

File: app/src/main/java/fr/free/nrw/commons/upload/UploadActivity.java
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/upload/UploadActivity.java:453]: javadoc

File: app/src/main/java/fr/free/nrw/commons/upload/UploadItem.java
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/upload/UploadItem.java:163]: javadoc

File: app/src/main/java/fr/free/nrw/commons/edit/EditViewModel.kt
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/edit/EditViewModel.kt:6]: class javadoc

File: app/src/main/java/fr/free/nrw/commons/edit/TransformImage.kt
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/edit/TransformImage.kt:5]: javadoc

File: app/src/main/java/fr/free/nrw/commons/edit/TransformImageImpl.kt
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/edit/TransformImageImpl.kt:11]: javadoc

File: app/src/main/java/fr/free/nrw/commons/edit/TransformImageImpl.kt
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/edit/TransformImageImpl.kt:12]: javadoc

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]

**Comment by shankarpriyank** [comment]
@nicolas-raoul I have added the java doc, I will investigate and report about the thumbnail issue by tomorrow

**Comment by nicolas-raoul** [comment]
Also, would you mind rebasing from the main branch? Thanks!

**Comment by shankarpriyank** [comment]
@nicolas-raoul Solving the thumbnail problem would take some time, I would create an issue for it once this pr is merged

**Comment by nicolas-raoul** [comment]
OK! 🙂

**Comment by shankarpriyank** [comment]
@nicolas-raoul did you get a chance to test this extensively?

**Comment by nicolas-raoul** [comment]
Yes! I uploaded about 100 pictures with your branch, including about 20 where I had to rotate. It worked great every time. :-)

File: app/src/main/java/fr/free/nrw/commons/upload/mediaDetails/UploadMediaDetailFragment.java
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/upload/mediaDetails/UploadMediaDetailFragment.java:574]: Would you mind using the same logging facility as the rest of the app?

File: app/src/main/java/fr/free/nrw/commons/upload/mediaDetails/UploadMediaDetailFragment.java
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/upload/mediaDetails/UploadMediaDetailFragment.java:567]: Could you please use the Guard Clause pattern here?
https://deviq.com/design-patterns/guard-clause
```
if (Objects.equals(result, "Error")) {
  log error and exit the block
}

try {
  etc
```
or similar

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]

**Comment by shankarpriyank** [comment]
Done @nicolas-raoul

File: app/src/main/java/fr/free/nrw/commons/edit/EditActivity.kt
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/edit/EditActivity.kt:81]: Ah, is all of this whitespace needed? :-)

File: app/src/main/java/fr/free/nrw/commons/edit/EditActivity.kt
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/edit/EditActivity.kt:162]: Single white lines is better within functions.

File: app/src/main/java/fr/free/nrw/commons/edit/EditActivity.kt
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/edit/EditActivity.kt:116]: remove empty line before }

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]
Sorry I just caught these.
Once the Flamingo PR is merged (now waiting for #5339), I will ask you to rebase (or pull), and will review the code once more just in case, thank you for your understanding. :-)

**Comment by shankarpriyank** [comment]
> Sorry I just caught these. Once the Flamingo PR is merged (now waiting for #5339), I will ask you to rebase (or pull), and will review the code once more just in case, thank you for your understanding. :-)

Hey! no worries at all.
Sorry for my formatting mistakes, I have fixed them now

File: app/src/main/java/fr/free/nrw/commons/upload/mediaDetails/UploadMediaDetailFragment.java
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/upload/mediaDetails/UploadMediaDetailFragment.java:585]: Please remove empty lines here too.

File: app/src/main/res/drawable/baseline_rotate_right.xml
Comment (nicolas-raoul) [inline app/src/main/res/drawable/baseline_rotate_right.xml:4]: Where did you get this icon from? If comments are allowed, a URL would be great for license tracking purposes.

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]

File: app/src/main/res/drawable/baseline_rotate_right.xml
Comment (shankarpriyank) [reply app/src/main/res/drawable/baseline_rotate_right.xml:4]: I got this icon from android studio itself, new -> vector asset -> clipart
The icon is licensed under [https://www.apache.org/licenses/LICENSE-2.0.txt](https://www.apache.org/licenses/LICENSE-2.0.txt)

Comment (nicolas-raoul) [reply app/src/main/res/drawable/baseline_rotate_right.xml:4]: Cool, I guess you can leave it like this then. Thanks!

**Comment by nicolas-raoul** [comment]
Actually, would you mind:
- From your same branch, creating a new pull request.
- On the [pull request creation webpage](https://user-images.githubusercontent.com/97883232/276500563-8d312351-b491-44e3-a877-2db3a6679431.png), select `base:v4.2.0-release` instead of `base:main` as the target branch.
- Send the pull request.

This will allow us to review all of your changes clearly (not mixed with other Flamingo-related changes). Maybe we can even merge this into the `v4.2.0-release` branch.

**Comment by shankarpriyank** [comment]
I made the pr, but in the first look it also seems to have unrelated changes.
I am a bit occupied for the next couples of days, will be having a closer look by monday

**Comment by nicolas-raoul** [comment]
The main branch now supports Flamingo :-)

**Comment by nicolas-raoul** [comment]
Would you mind rebasing or pulling from main, then checking the difference and possibly donig a bit of clean-up? Then ask for a review. :-)

**Comment by nicolas-raoul** [comment]
Any idea why a unit test fails?

**Comment by shankarpriyank** [comment]
> Any idea why a unit test fails?

Nope, had a look at the logs, I could not really infer much from it apart from

> Caused by: java.lang.IllegalStateException: Cannot process instrumented class fr/free/nrw/commons/di/DaggerCommonsApplicationComponent$60. Please supply original non-instrumented classes.

**Comment by nicolas-raoul** [comment]
```
fr.free.nrw.commons.settings.SettingsActivityUnitTests > testSetTotalUploadCount FAILED
    java.lang.reflect.InvocationTargetException at SettingsActivityUnitTests.kt:77
        Caused by: java.lang.IllegalStateException at SettingsActivityUnitTests.kt:77
```

**Comment by shankarpriyank** [comment]
@nicolas-raoul its fixed now

File: app/build.gradle
Comment (nicolas-raoul) [inline app/build.gradle:240]: I think we should keep test coverage, though?

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]

File: app/build.gradle
Comment (nicolas-raoul) [reply app/build.gradle:240]: `SettingsActivityUnitTests` failures does not seem to happen on main, somehow (it fails with other errors though): https://github.com/commons-app/apps-android-commons/actions/runs/6583628746/job/17886967295

Comment (shankarpriyank) [reply app/build.gradle:240]: Fixed

File: app/src/main/java/fr/free/nrw/commons/upload/mediaDetails/UploadMediaDetailFragment.java
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/upload/mediaDetails/UploadMediaDetailFragment.java:482]: javadoc

File: app/src/main/java/fr/free/nrw/commons/upload/mediaDetails/UploadMediaDetailFragment.java
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/upload/mediaDetails/UploadMediaDetailFragment.java:488]: remove 2 empty lines

File: app/src/main/java/fr/free/nrw/commons/upload/mediaDetails/UploadMediaPresenter.java
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/upload/mediaDetails/UploadMediaPresenter.java:279]: remove 1 empty line

File: app/src/main/java/fr/free/nrw/commons/edit/EditActivity.kt
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/edit/EditActivity.kt:79]: remove 1 empty line

File: app/src/main/java/fr/free/nrw/commons/edit/EditActivity.kt
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/edit/EditActivity.kt:40]: insert 1 empty line before function

File: app/src/main/java/fr/free/nrw/commons/edit/EditActivity.kt
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/edit/EditActivity.kt:200]: remove 1 empty line

File: app/src/main/java/fr/free/nrw/commons/edit/EditActivity.kt
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/edit/EditActivity.kt:229]: remove 1 empty line

File: app/src/main/java/fr/free/nrw/commons/edit/EditViewModel.kt
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/edit/EditViewModel.kt:26]: remove 1 empty line

File: app/src/main/java/fr/free/nrw/commons/edit/TransformImageImpl.kt
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/edit/TransformImageImpl.kt:74]: remove 1 empty line

File: app/src/main/java/fr/free/nrw/commons/upload/UploadItem.java
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/upload/UploadItem.java:175]: remove 2 empty lines

File: app/src/main/java/fr/free/nrw/commons/upload/mediaDetails/UploadMediaDetailFragment.java
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/upload/mediaDetails/UploadMediaDetailFragment.java:70]: remove 1 empty line

File: app/src/main/java/fr/free/nrw/commons/upload/mediaDetails/UploadMediaDetailFragment.java
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/upload/mediaDetails/UploadMediaDetailFragment.java:279]: remove 1 empty line

File: app/src/main/java/fr/free/nrw/commons/upload/mediaDetails/UploadMediaDetailFragment.java
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/upload/mediaDetails/UploadMediaDetailFragment.java:281]: remove 1 empty line

File: app/src/main/java/fr/free/nrw/commons/upload/mediaDetails/UploadMediaDetailFragment.java
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/upload/mediaDetails/UploadMediaDetailFragment.java:590]: remove 1 empty line

File: app/src/main/java/fr/free/nrw/commons/upload/mediaDetails/UploadMediaDetailsContract.java
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/upload/mediaDetails/UploadMediaDetailsContract.java:62]: only 1 empty line

File: app/src/main/res/layout/activity_edit.xml
Comment (nicolas-raoul) [inline app/src/main/res/layout/activity_edit.xml:46]: remove 2 empty lines

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]
minor

**Comment by shankarpriyank** [comment]
@nicolas-raoul i guess all of them are fixed now

File: app/src/main/java/fr/free/nrw/commons/upload/mediaDetails/UploadMediaDetailFragment.java
Comment (nicolas-raoul) [reply app/src/main/java/fr/free/nrw/commons/upload/mediaDetails/UploadMediaDetailFragment.java:477]: Even for overridden methods, it is good to write javadoc (which can be taken from the parent class if applicable). Every written javadoc is then read by hundreds of people, so having the information wherever it is needed is important.

File: app/src/main/java/fr/free/nrw/commons/upload/mediaDetails/UploadMediaDetailFragment.java
Comment (nicolas-raoul) [reply app/src/main/java/fr/free/nrw/commons/upload/mediaDetails/UploadMediaDetailFragment.java:581]: remove 1 line

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]

**Comment by shankarpriyank** [comment]
fixed

**Comment by nicolas-raoul** [comment]
Huge congratulations @shankarpriyank for adding this lossless rotation feature to the app, as well as setting the foundation for other lossless transforms.
Thanks to you, many picture uploads will have both keep their original quality and not occupy more disk space, while being in the right orientation to be used in Wikipedia articles and other places.
You truly help Wikipedia/etc run better!

**Comment by shankarpriyank** [comment]
Thank you so much for your kind words!

**Pull request merged**
